File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/MVVSymbol.java
package de.tum.in.tumcampusapp.component.ui.transportation;

import android.graphics.Color;

import com.google.common.base.Optional;
import com.google.common.primitives.Ints;

/**
 * Encapsulates information about the symbol shown next to departure information. Contains the
 * background color and text color, as computed based on the line type and line number.
 */
public class MVVSymbol {

    private static final int[] S_LINE_COLOR = {
            0xff6db9df, 0xff8db335, 0xff7d187b, 0xffc1003c, 0, 0xff00915e, 0xff7f3229, 0xff1f1e21
    };

    private static final int[] U_LINE_COLOR = {
            0xff44723c, 0xffcc263c, 0xffe4721c, 0xff04a27c, 0xffa4721c, 0xff045ea4, 0xffc10134, 0xffeb6a27
    };

    private static final int U_LINE_DEFAULT_COLOR = Color.GRAY;

    private final int mBackgroundColor;
    private final int mTextColor;

    /**
     * Standard constructor
     *
     * @param line Line symbol name e.g. U6, S1, T14
     */
    public MVVSymbol(String line) {
        int textColor = 0xffffffff;
        int backgroundColor = 0;

        char symbol = !line.isEmpty() ? line.charAt(0) : 'X';
        Integer parsedNumber = Ints.tryParse(line.substring(1));
        int lineNumber = Optional.fromNullable(parsedNumber).or(0);

        switch (symbol) {
            case 'S':
                if (lineNumber > 0 && lineNumber <= 8) {
                    backgroundColor = S_LINE_COLOR[lineNumber - 1];
                } else if (lineNumber == 20) {
                    backgroundColor = 0xffca536a;
                } else if (lineNumber == 27) {
                    backgroundColor = 0xffd99098;
                }
                textColor = lineNumber == 8 ? 0xfff1cb00 : 0xffffffff;
                break;
            case 'U':
                if (lineNumber > 0 && lineNumber <= U_LINE_COLOR.length) {
                    backgroundColor = U_LINE_COLOR[lineNumber - 1];
                } else {
                    backgroundColor = U_LINE_DEFAULT_COLOR;
                }
                textColor = 0xfffcfefc;
                break;
            case 'N':
                backgroundColor = 0xff000000;
                textColor = 0xffebd22e;
                break;
            case 'X':
                backgroundColor = 0xff477f70;
                break;
            default:
                if (lineNumber < 50) {
                    backgroundColor = 0xffdc261c;
                    textColor = 0xfffcfefc;
                } else if (lineNumber < 90) {
                    backgroundColor = 0xfffc6604;
                } else {
                    backgroundColor = 0xff004a5d;
                }
                break;
        }

        mTextColor = textColor;
        mBackgroundColor = backgroundColor;
    }

    public int getTextColor() {
        return mTextColor;
    }

    public int getBackgroundColor() {
        return mBackgroundColor;
    }

}

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/MVVSymbol.kt
package de.tum.`in`.tumcampusapp.component.ui.transportation

import android.content.Context
import de.tum.`in`.tumcampusapp.R

/**
 * Encapsulates information about the symbol shown next to departure information. Contains the
 * background color and text color, as computed based on the line type and line number.
 *
 * @param line Line symbol name e.g. U6, S1, T14
 */
class MVVSymbol(line: String, val context: Context) {

    val backgroundColor: Int
    val textColor: Int

    init {
        var textColor = R.color.white
        val backgroundColor: Int

        val symbol = line.getOrNull(0) ?: 'X'
        val lineNumber = line.substring(1).toIntOrNull() ?: 0

        when (symbol) {
            'S' -> {
                backgroundColor = when (lineNumber) {
                    in 1..8 -> S_LINE_COLOR[lineNumber - 1]
                    20 -> R.color.s_20
                    else -> R.color.unknown_line
                }
                textColor = if (lineNumber == 8) R.color.s_8_font else R.color.white
            }
            'U' -> {
                backgroundColor = if (lineNumber > 0 && lineNumber <= U_LINE_COLOR.size) {
                    U_LINE_COLOR[lineNumber - 1]
                } else {
                    R.color.unknown_line
                }
                textColor = when (lineNumber) {
                    7 -> R.color.u_7_accent
                    8 -> R.color.u_8_accent
                    else -> R.color.white
                }
            }
            'N' -> {
                backgroundColor = R.color.black
                textColor = R.color.night_line_font
            }
            'X' -> backgroundColor = R.color.express_bus
            else -> backgroundColor = when {
                lineNumber < 50 -> R.color.tram
                lineNumber < 90 -> R.color.metro_bus
                else -> R.color.regular_bus
            }
        }

        this.textColor = context.resources.getColor(textColor)
        this.backgroundColor = context.resources.getColor(backgroundColor)
    }

    fun getHighlight(): Int {
        return context.resources.getColor(R.color.reduced_opacity) and backgroundColor
    }

    companion object {
        private val S_LINE_COLOR = intArrayOf(
                R.color.s_1,
                R.color.s_2,
                R.color.s_3,
                R.color.s_4,
                R.color.unknown_line,
                R.color.s_6,
                R.color.s_7,
                R.color.s_8)

        private val U_LINE_COLOR = intArrayOf(
                R.color.u_1,
                R.color.u_2,
                R.color.u_3,
                R.color.u_4,
                R.color.u_5,
                R.color.u_6,
                R.color.u_7,
                R.color.u_8)
    }
}
